### PR TITLE
Fix for missing callstacks in real time Tracelog sessions (issue 332)

### PR DESCRIPTION
**Pull request by Alois-xx**
When using  Realtime tracing the call stacks of x64 processes did come up empty because the code addresses were treated as kernel addresses which did result in the assignment of 0 (IDLE) process instead of the real processes where the stack was generated.  See issue #332

### CODE REVIEW

File: src/TraceEvent/ETWReloggerTraceEventSource.cs [hunk 1]
```diff
@@ -407,7 +407,7 @@ private unsafe void Initialize(TraceEventNativeMethods.EVENT_RECORD* rawData)
                         // first event (real time case).   This is really a problem, as we need that information, but we will go ahead and
                         // try to initialize as best we can. 
 
-                        m_source.pointerSize = sizeof(IntPtr);
+                        m_source.pointerSize = Environment.Is64BitOperatingSystem ? 8 : 4;
```
Comment (jorive) [inline src/TraceEvent/ETWReloggerTraceEventSource.cs]: Is it possible to use `IntPtr.Size` instead?

File: src/TraceEvent/ETWTraceEventSource.cs [hunk 2]
```diff
@@ -455,9 +455,9 @@ private void Initialize(string fileOrSessionName, TraceEventSourceType type)
                 sessionEndTimeQPC = this.UTCDateTimeToQPC(maxSessionEndTimeUTC);
             }
             Debug.Assert(_QPCFreq != 0);
-            if (pointerSize == 0)       // Real time does not set this (grrr). 
+            if (pointerSize == 0 || IsRealTime)  // We get on x64 OS 4 as pointer size which is wrong for realtime sessions. Fix it up. 
             {
-                pointerSize = sizeof(IntPtr);
+                pointerSize = Environment.Is64BitOperatingSystem ? 8 : 4;
```
Comment (jorive) [inline src/TraceEvent/ETWTraceEventSource.cs]: Is it possible to use IntPtr.Size instead?

File: src/TraceEvent/ETWReloggerTraceEventSource.cs (hunk 1 above)
Comment (Alois-xx) [inline src/TraceEvent/ETWReloggerTraceEventSource.cs]: I am not sure what is right here. When an ETL is recorded it can contain x64 and x86 stacks. So the structures in the ETL file should be all dependant on the bitness of the OS and not the process starting the trace. 
It could be that depending on the source.pointersize various assumptions are encoded which will lead to followup errors. 
I can strip the change down to the concrete fix to make realtime tracing work and leave the rest as is.

Comment (vancem) [inline src/TraceEvent/ETWReloggerTraceEventSource.cs]: I agree that pointerSize should be the underlying OS pointer size (we should never be using IntPtr (that is the size for the local process) to determine this value.    This is a good change.

File: src/TraceEvent/ETWTraceEventSource.cs [hunk 3]
```diff
@@ -455,9 +455,9 @@ private void Initialize(string fileOrSessionName, TraceEventSourceType type)
                 sessionEndTimeQPC = this.UTCDateTimeToQPC(maxSessionEndTimeUTC);
             }
             Debug.Assert(_QPCFreq != 0);
-            if (pointerSize == 0)       // Real time does not set this (grrr). 
+            if (pointerSize == 0 || IsRealTime)  // We get on x64 OS 4 as pointer size which is wrong for realtime sessions. Fix it up.
```
Comment (vancem) [inline src/TraceEvent/ETWTraceEventSource.cs]: Ideally we would not have the || IsRealTime in this condition.   Basically the ETW system in theory knows the pointer size and we should just be using that.  We only 'take control' and set it if it is clear that the ETW system did not (it set it to 0).   

Is it really the case that the pointerSize comes back as 4 on a when you create a real-time ETW source?

File: src/TraceEvent/NativeDlls.cs [hunk 4]
```diff
@@ -15,7 +15,9 @@ class NativeDlls
     /// <param name="simpleName"></param>
     public static void LoadNative(string simpleName)
     {
-        var thisDllDir = Path.GetDirectoryName(Assembly.GetExecutingAssembly().ManifestModule.FullyQualifiedName);
+        // When TraceEvent is loaded as embedded assembly the executing assembly is mscorlib which points to the wrong path.
```
Comment (vancem) [inline src/TraceEvent/NativeDlls.cs]: I would like to understand this scenario more.   GetExeuctingAssembly gets the assembly of the CALLER of that API (otherwise it would be useless, as you point out it would always return mscorlib).   

Thus the existing code will look for the DLL 'next to' the assembly that compiled NativeDlls.cs (which would be Microsoft.Diagnostics.Tracing.TraceEvent.dll).  That is what I want.   There are scenarios where you don't control
the entry assembly (that is the EXE) so you don't really want to use that.   

Basically I am asking for more details about the issue that drove this change.   I think the code as it was should work.

File: src/TraceEvent/ETWTraceEventSource.cs (hunk 3 above)
Comment (Alois-xx) [inline src/TraceEvent/ETWTraceEventSource.cs]: Yes I have checked that on x64 I have got 4 as pointer size back. I guess that can happen if you start tracing from an x86 process but I have not looked deeper but the  set value was definitely wrong.

File: src/TraceEvent/NativeDlls.cs (hunk 4 above)
Comment (Alois-xx) [inline src/TraceEvent/NativeDlls.cs]: I have several tools which rely on TraceEvent in different versions. I would like to embed TraceEvent and related things and load them as byte arrays into the AppDomain so I do not need to deploy the referenced assemblies to prevent breaking any tool relying on TraceEvent. 

I just have found that OSExtensions.dll
`private static void LoadKernelTraceControl()
{
	if (!ETWKernelControl.s_KernelTraceControlLoaded)
	{
		try
		{
			string arg_27_0 = Path.GetDirectoryName(Assembly.GetExecutingAssembly().ManifestModule.FullyQualifiedName);
`

uses the same trick which breaks the approach. I know that these wrapper are searched relative to the dll but if it is loaded from a byte array the Manifest path will be <Unknown> which then breaks the Path.Combine call due to invalid characters. 
A non breaking workaround would be to check first the returned ManifestModule.FullyQualifiedName for <Unkown>. If this is bad then use the executable directory as fallback in TraceEvent and OsExtensions. That way the old deployments are not broken and new ones are enabled.

Comment (Alois-xx) [inline src/TraceEvent/NativeDlls.cs]: By the way the publicly used KernelTraceControl.dll does not work on Windows 7 because it is linked with ApiSets agains Win8? It fails with  

`Application popup: LiveExceptionMonitor.exe - Entry Point Not Found : The procedure entry point ReleaseSRWLockExclusive could not be located in the dynamic link library api-ms-win-core-synch-l1-2-0.dll. 
`
That should be fixed as well. 
As feature request I would like to use "Circular Kernel Context Logger" as my second trace logger to get out of the way of procmon, procexp, ... and "real" profiling. That would also be a big help. This is a big help on Win7 where officially there is only one kernel logger allowed.
